server/src/main/java/org/elasticsearch/cluster/coordination/Coordinator.java

```diff
@@ -366,6 +401,242 @@ public ClusterState getLastAcceptedState() {
+        synchronized (mutex) {
+            // expose last accepted cluster state as base state upon which the master service
+            // speculatively calculates the next cluster state update
+            final ClusterState clusterState = coordinationState.get().getLastAcceptedState();
+            if (mode != Mode.LEADER || clusterState.term() != getCurrentTerm()) {
+                // the master service checks if the local node is the master node in order to fail execution of the state update early
+                return clusterStateWithNoMasterBlock(clusterState);
+            }
+            return clusterState;
+        }
+    }
+
+    private ClusterState clusterStateWithNoMasterBlock(ClusterState clusterState) {
+        if (clusterState.nodes().getMasterNodeId() != null) {
+            // remove block if it already exists before adding new one
+            assert clusterState.blocks().hasGlobalBlock(DiscoverySettings.NO_MASTER_BLOCK_ID) == false :
+                "NO_MASTER_BLOCK should only be added by Coordinator";
+            // TODO: allow dynamically configuring NO_MASTER_BLOCK_ALL
+            final ClusterBlocks clusterBlocks = ClusterBlocks.builder().blocks(clusterState.blocks()).addGlobalBlock(
+                DiscoverySettings.NO_MASTER_BLOCK_WRITES).build();
+            final DiscoveryNodes discoveryNodes = new DiscoveryNodes.Builder(clusterState.nodes()).masterNodeId(null).build();
+            return ClusterState.builder(clusterState).blocks(clusterBlocks).nodes(discoveryNodes).build();
+        } else {
+            return clusterState;
+        }
+    }
+
+    @Override
+    public void publish(ClusterChangedEvent clusterChangedEvent, ActionListener<Void> publishListener, AckListener ackListener) {
+        try {
+            synchronized (mutex) {
+                publishUnderLock(clusterChangedEvent, publishListener, ackListener);
+            }
+        } catch (Exception e) {
+            logger.trace(() -> new ParameterizedMessage("[{}] publishing: [{}] failed: {}",
+                getLocalNode().getName(), clusterChangedEvent.source(), e.getMessage()), e);
+            publishListener.onFailure(new FailedToCommitClusterStateException("failure while publishing", e));
+        }
+    }
+
+    private void publishUnderLock(ClusterChangedEvent clusterChangedEvent, ActionListener<Void> completionListener,
+                                  AckListener ackListener) {
+        assert Thread.holdsLock(mutex) : "Coordinator mutex not held";
+
+        if (mode != Mode.LEADER) {
+            throw new CoordinationStateRejectedException("publishUnderLock: not currently leading, so cannot handle client value.");
+        }
+
+        if (currentPublication.isPresent()) {
+            throw new CoordinationStateRejectedException("[{}] is in progress", currentPublication.get());
+        }
+
+        // the DiscoveryNodes part can be different as we adapt it in getStateForMasterService when becoming master
```

Comment:
I tried turning this comment into an assertion which overwrites the `nodes()` of `clusterChangedEvent.previousState()` with those from `coordinationState.get().getLastAcceptedState()` and then asserts it to be equal to `coordinationState.get().getLastAcceptedState()` and discovered that there's no `.equals` method on `ClusterState` so it didn't work. But it'd be useful if we could do this.

Reply:
I've reworked the assertion in ab43fb15493. Let me know if you like it better that way.